### ISSUE TITLE
cmd/go: malformed go.sum error is ignored

### ISSUE THREAD

**Issue by chmeliik** [golang/go#62345]
### What version of Go are you using (`go version`)?

<pre>
$ go version
go version go1.19.12 linux/amd64
</pre>

### Does this issue reproduce with the latest release?

yes

### What operating system and processor architecture are you using (`go env`)?

<details><summary><code>go env</code> Output</summary><br><pre>
$ go env
GO111MODULE=""
GOARCH="amd64"
GOBIN=""
GOCACHE="/home/acmiel/.cache/go-build"
GOENV="/home/acmiel/.config/go/env"
GOEXE=""
GOEXPERIMENT=""
GOFLAGS=""
GOHOSTARCH="amd64"
GOHOSTOS="linux"
GOINSECURE=""
GOMODCACHE="/home/acmiel/go/pkg/mod"
GONOPROXY=""
GONOSUMDB=""
GOOS="linux"
GOPATH="/home/acmiel/go"
GOPRIVATE=""
GOPROXY="https://proxy.golang.org,direct"
GOROOT="/usr/lib/golang"
GOSUMDB="off"
GOTMPDIR=""
GOTOOLDIR="/usr/lib/golang/pkg/tool/linux_amd64"
GOVCS=""
GOVERSION="go1.19.12"
GCCGO="gccgo"
GOAMD64="v1"
AR="ar"
CC="gcc"
CXX="g++"
CGO_ENABLED="1"
GOMOD="/tmp/test/go.mod"
GOWORK=""
CGO_CFLAGS="-g -O2"
CGO_CPPFLAGS=""
CGO_CXXFLAGS="-g -O2"
CGO_FFLAGS="-g -O2"
CGO_LDFLAGS="-g -O2"
PKG_CONFIG="pkg-config"
GOGCCFLAGS="-fPIC -m64 -pthread -Wl,--no-gc-sections -fmessage-length=0 -fdebug-prefix-map=/tmp/go-build2703022439=/tmp/go-build -gno-record-gcc-switches"
</pre></details>

### What did you do?

Intentionally break a go.sum file and observe `go mod download` behavior

<details>
<summary>create test project</summary>

```shell
cat << EOF > main.go
package main

import (
	"fmt"

	"rsc.io/quote/v4"
)

func main() {
	fmt.Println(quote.Hello())
}
EOF

cat << EOF > go.mod
module acmiel.test/foo

go 1.19

require rsc.io/quote/v4 v4.0.1

require (
	golang.org/x/text v0.0.0-20170915032832-14c0d48ead0c // indirect
	rsc.io/sampler v1.3.0 // indirect
)
EOF

cat << EOF > go.sum
golang.org/x/text v0.0.0-20170915032832-14c0d48ead0c h1:qgOY6WgZOaTkIIMiVjBQcw93ERBE4m30iBm00nkL0i8=
golang.org/x/text v0.0.0-20170915032832-14c0d48ead0c/go.mod h1:NqM8EUOU14njkJ3fqMW+pc6Ldnwhi/IjpwHt7yyuwOQ=
rsc.io/quote/v4 v4.0.1
rsc.io/quote/v4 v4.0.1/go.mod h1:w/DafQky66grMesu3uPhdDMS3knhBippwwemZtMOyCI=
rsc.io/sampler v1.3.0 h1:7uVkIFmeBqHfdjD+gZwtXXI+RODJ2Wc4O7MPEh/QiW4=
rsc.io/sampler v1.3.0/go.mod h1:T1hPZKmBbMNahiBKFy5HrXp6adAjACjK9JXDnKaTXpA=
EOF
```

</details>

go.sum (note that line 3 is broken):

```go.sum
golang.org/x/text v0.0.0-20170915032832-14c0d48ead0c h1:qgOY6WgZOaTkIIMiVjBQcw93ERBE4m30iBm00nkL0i8=
golang.org/x/text v0.0.0-20170915032832-14c0d48ead0c/go.mod h1:NqM8EUOU14njkJ3fqMW+pc6Ldnwhi/IjpwHt7yyuwOQ=
rsc.io/quote/v4 v4.0.1
rsc.io/quote/v4 v4.0.1/go.mod h1:w/DafQky66grMesu3uPhdDMS3knhBippwwemZtMOyCI=
rsc.io/sampler v1.3.0 h1:7uVkIFmeBqHfdjD+gZwtXXI+RODJ2Wc4O7MPEh/QiW4=
rsc.io/sampler v1.3.0/go.mod h1:T1hPZKmBbMNahiBKFy5HrXp6adAjACjK9JXDnKaTXpA=
```

go.sum after running `go mod download`:

```go.sum
golang.org/x/text v0.0.0-20170915032832-14c0d48ead0c h1:qgOY6WgZOaTkIIMiVjBQcw93ERBE4m30iBm00nkL0i8=
golang.org/x/text v0.0.0-20170915032832-14c0d48ead0c/go.mod h1:NqM8EUOU14njkJ3fqMW+pc6Ldnwhi/IjpwHt7yyuwOQ=
rsc.io/quote/v4 v4.0.1/go.mod h1:w/DafQky66grMesu3uPhdDMS3knhBippwwemZtMOyCI=
rsc.io/sampler v1.3.0/go.mod h1:T1hPZKmBbMNahiBKFy5HrXp6adAjACjK9JXDnKaTXpA=
```

The zip checksums in the broken part of go.sum were removed. I think the reason is that readGoSum [returns an error](https://github.com/golang/go/blob/eaae2d45c76d2d27e1b4c39ba57dd1582ab26491/src/cmd/go/internal/modfetch/fetch.go#L523-L525) but the callers of readGoSum ([first](https://github.com/golang/go/blob/eaae2d45c76d2d27e1b4c39ba57dd1582ab26491/src/cmd/go/internal/modfetch/fetch.go#L495), [second](https://github.com/golang/go/blob/eaae2d45c76d2d27e1b4c39ba57dd1582ab26491/src/cmd/go/internal/modfetch/fetch.go#L836)) ignore that error. Effectively, this means that commands which parse go.sum will parse only the non-broken part and ignore the rest.

### What did you expect to see?

An error telling me that my go.sum is broken

### What did you see instead?

No error, handling only the non-broken part of go.sum

**Comment by bcmills** [comment]
I agree that `go mod tidy` should continue to fix a broken `go.sum` file, and I also agree that filling in those missing error checks seems like a good idea. Want to send a fix?

**Comment by chmeliik** [comment]
@bcmills I'd be happy to!

**Comment by chmeliik** [comment]
Ended up not being as simple as I expected (simply filling in the missing error checks did not achieve what I needed)

Nevertheless, I've sent a somewhat hacky fix :sweat_smile:
